### ISSUE TITLE
Dictionary lookup stops working after a while

### ISSUE THREAD

**Issue by nairyosangha** [koreader/koreader#10312]
* KOReader version: 2023.03
* Device: kobo aura 4.21.15015

#### Issue
Ever since upgrading to 2023.03 I started having strange issues while looking up words.
When I start the reader, everything goes fine, but as I start doing lookups, it feels like the reader gets slower and slower, and eventually it gets stuck on the 'Searching dictionary for: word' message.
I can still click this message away, but at this point it won't ever do a successful lookup anymore, until I restart KOreader.
Since the responsiveness of the whole reader also starts feeling more and more sluggish as time passes, I figured it might be some memory related issue:

when I start up the reader:
```sh
Every 5s: free -m                                                                                                                                                                                                                                    2023-04-10 11:37:45

             total         used         free       shared      buffers
Mem:           247           68          178            0            1
-/+ buffers:                 66          180
Swap:            0            0            0
```
with each new lookup, the free memory goes down, eventually bottoming out at
```sh
Every 5s: free -m                                                                                                                                                                                                                                    2023-04-10 11:37:10

             total         used         free       shared      buffers
Mem:           247          237           10            0            0
-/+ buffers:                237           10
Swap:            0            0            0
```

I can reproduce this reliably, although sometimes I need to do substantially more lookups to get in this state, so I'm not sure if that's what's causing this exactly.

This memory usage might not actually be the problem though.
I noticed that getting rid of the custom CSS I have for one of my dictionaries makes the issue go away. I had this CSS already before I updated.
```css
@font-face {
	font-family: "Noto Sans CJK JP";
	src: url("/mnt/onboard/.adds/koreader/data/dict/(EDRDG) JMdict [Version 2023.01.03]/data/NotoSansCJKjp-Regular.otf");
	font-weight: "regular";
	font-style: "normal";
}
big {
	font-family: "Noto Sans CJK JP";
}
```
The used memory still jumps up by a huge amount on the first lookup, and it goes down to about the same amount free, but it starts fluctuating then, and it doesn't seem to get stuck anymore. 
#### Steps to reproduce

##### `crash.log` (if applicable)
`crash.log` is a file that is automatically created when KOReader crashes. It can normally be found in the KOReader directory:

[crash.log](https://github.com/koreader/koreader/files/11189981/crash.log)



Am I doing something obviously wrong in that CSS? It is loading a font that's 17MB, so that might be an issue. I don't understand why I suddenly started having the issue only after bumping versions though..

Thanks in advance!

**Comment by Frenzie** [comment]
You could try inserting a `collectgarbage()` (or two) at the end of closing the dictionary display though I can't see how anything could get stuck from gc regardless.

**Comment by NiLuJe** [comment]
If there are HTML dictionaries involved, I could certainly see MuPDF shenanigans happening that would at the very least commit a certain amount of memory (if not outright leaking some).

**Comment by nairyosangha** [comment]
I'm sure I upgraded from 2022.11, however I put that release on my ereader, and I can still reproduce it :sweat_smile: 

I only started messing with that font stuff a month or so ago, so I guess it's not actually related to the new version, I don't see big changes in those releases for mupdf either

**Comment by NiLuJe** [comment]
Yeah, the font certainly looks like a good suspect ;).

(I don't recall *at all* how this stuff works in MuPDF).

**Comment by nairyosangha** [comment]
When I disable the 1 HTML dict I have completely, memory usage seems completely stable while reading/doing lookups.

When turning it on, but without custom CSS, memory usage drops substantially, goes down to low 10s free, but it seems to 'recover' quickly enough to never cause an issue.

When I have the custom CSS which loads the font, free memory drops WAY faster, and I hit the issue pretty quickly.

**Comment by nairyosangha** [comment]
Tried going about it another way, looking back at what @poire-z told me on gitter before:

> but from your screenshot, as it has italic, it is not a text dict, but an HTML dict - and there, it's not our UI text code (that I described above) that is used, but MuPDF that is used to render the html in the dictresult window - and we have little control and knowledge on how it renders it - so it might use NotoSansCJK, but with its default Chinese variants if 1) MuPDF does not support and <span lang=JA> in the HTML output or 2) there is no such \<span lang=JA\>

and the mupdf code: https://github.com/ArtifexSoftware/mupdf/blob/4a7822d6750ecb6e0b63f4357738dc20ecaa58f6/source/html/html-layout.c#L736

I tried adding a custom lua file in my dict folder instead of custom css, like so
```lua
return function(html)
    html = "<div lang=\"ja\">"..html.."</div>"
    return html
end
```

and what do you know, that actually works :smile: , doesn't have the annoying lag either

**Comment by Frenzie** [comment]
Sidenote but I tend to prefer using `'<div lang="ja">'` for cases like that. (Or `[[]]`.)

**Comment by nairyosangha** [comment]
> Sidenote but I tend to prefer using `'<div lang="ja">'` for cases like that. (Or `[[]]`.)

that does look nicer, will go with that :slightly_smiling_face:

**Comment by poire-z** [comment]
It's possible there is a memory leak in the old MuPDF version we are still using, when it has to handle `@font-face { src: url(...) }`, like not releasing the font object, so the more lookups you do, the more duplicates of that font object are kept in memory....
(MuPdf is a bit of a black box to me, so not gonna investigate :/)

**Comment by Frenzie** [comment]
In any case, I wouldn't want to investigate MuPDF issues too hard while we're on the older version. :-)
